File: jammy_app/jammy_app/custom_jammy/delivery_note/delivery_note.js
cur_frm.add_fetch('warehouse','address_title','address_title');
cur_frm.add_fetch('warehouse','address_line1','address_line1');
cur_frm.add_fetch('warehouse','address_line2','address_line2');
cur_frm.add_fetch('warehouse','city','city');
cur_frm.add_fetch('warehouse','state','state');
cur_frm.add_fetch('warehouse','pincode','pincode');
// cur_frm.add_fetch('customer','credit_days_based_on','credit_days_based_on');
// cur_frm.add_fetch('customer','credit_days','credit_days');
// cur_frm.add_fetch('customer','credit_limit','credit_limit');
cur_frm.add_fetch('customer','payment_terms','payment_terms');
cur_frm.add_fetch('customer','freight_ppd','freight_ppd');
cur_frm.add_fetch('customer','shipping_account','shipping_account');
cur_frm.add_fetch('customer','email_option','email_option');
// "ignore_pricing_rule : 1",
cur_frm.add_fetch('item_code', 'net_weight', 'net_weight');
cur_frm.add_fetch('item_code', 'pcs_ctn', 'pcs_ctn');

frappe.ui.form.on("Delivery Note Item", "cartons", function(frm) {
	set_total_cartons(frm);
});

// frappe.ui.form.on("Delivery Note Item", "cartons", function(frm, cdt, cdn){
// 	var items = frm.doc.items
// 	let total_cartons =0;
// 	for(var i in items) {
// 		total_cartons = total_cartons + items[i].cartons
// 	}
//    	frm.set_value('total_cartons', total_cartons);
//    	frm.set_df_property("total_cartons", "read_only", 1);
// });

frappe.ui.form.on("Delivery Note Item", "groos_weight", function(frm) {
	set_total_weight(frm);
});

// frappe.ui.form.on("Delivery Note Item", "gross_weight", function(frm, cdt, cdn){
// 	var items = frm.doc.items
// 	let total_weight=0;
// 	for(var i in items) {
// 		total_weight = total_weight + items[i].gross_weight
// 	}
//    	frm.set_value('total_weight_pkg', total_weight);
//    	frm.set_df_property("total_weight_pkg", "read_only", 1);
// });

frappe.ui.form.on("Delivery Note Item", "qty", function(frm, cdt, cdn) {
	set_cartons(frm, cdt, cdn);
	set_gross_weight(frm, cdt, cdn);
	set_total_cartons(frm);
	set_total_weight(frm);
});
frappe.ui.form.on("Delivery Note Item", "net_weight", function(frm, cdt, cdn) {
	set_gross_weight(frm, cdt, cdn);
	set_total_weight(frm);
});

frappe.ui.form.on("Delivery Note Item", "pcs_ctn", function(frm, cdt, cdn) {
	set_cartons(frm, cdt, cdn);
	set_total_cartons(frm);
});

var set_cartons = function(frm, cdt, cdn) {
	var row = locals[cdt][cdn];
	if (row.pcs_ctn) {
		var qty = row.qty || 1;
		var cartons = Math.ceil(flt(qty) / flt(row.pcs_ctn));
		frappe.model.set_value(cdt, cdn, "cartons", cartons);
	}
}
var set_gross_weight = function(frm, cdt, cdn) {
	var row = locals[cdt][cdn];
	if (row.net_weight) {
		var qty = row.qty || 1;
		var gross_weight = flt(qty) * flt(row.net_weight);
		frappe.model.set_value(cdt, cdn, "gross_weight", gross_weight);
	}
}

var set_total_cartons = function(frm) {
	let sum_cartons = 0;
	$.each(frm.doc.items, function(i, d) {
		sum_cartons += d.cartons;
	});
	frm.set_value("total_cartons", sum_cartons);
	frm.set_df_property("total_cartons", "read_only", 1);
}
var set_total_weight = function(frm) {
	let  total_weight = 0;
	$.each(frm.doc.items, function(i, d) {
		total_weight += d.gross_weight;
	});
	frm.set_value("total_weight_pkg", total_weight);
	frm.set_df_property("total_weight_pkg", "read_only", 1);
}

frappe.ui.form.on("Delivery Note", {
	refresh: function (frm) {
		frm.add_custom_button(
			"Bill Of Lading",
			function () {
				let dialog = new frappe.ui.Dialog({
					title: 'Choose Type of BOL',
					fields: [{
						label: 'Choose Type of BOL',
						fieldname: 'radio_button_container',
						fieldtype: 'HTML',
						options: `<fieldset><div><input type="radio" id="direct_item_info" name="type_of_bol" value="direct_item_info" checked><label for="direct_item_info">Direct DN Item info</label></div><div><input type="radio" id="one_line_bol" name="type_of_bol" value="one_line_bol"><label for="one_line_bol">One line BOL item</label></div><div><input type="radio" id="normal_bol" name="type_of_bol" value="normal_bol"><label for="normal_bol">Normal BOL(aggregate by item group)</label></div></fieldset>`
					}, ],
					size: 'small',
					primary_action_label: 'Create BOL',
					primary_action(values) {
						let type_of_bol = $('input[name="type_of_bol"]:checked').val();
						console.log(type_of_bol)
						dialog.fields_dict.radio_button_container.$wrapper.empty();
						dialog.hide();
						frappe.call({
							method: "jammy_app.jammy_app.custom_jammy.delivery_note.delivery_note.create_bol",
							args: {
								'doctype': 'Delivery Note',
								'docname': frm.doc.name,
								'type_of_bol': type_of_bol,
							},
							callback: function (r) {
								if (!r.exc) {
									frappe.msgprint('Bill of Lading created successfully.');
								}
							},
							success: function (r) {},
							url: "" || frappe.request.url,
						});
					}
				});
				dialog.show();
			}, "Create");
	}
})

// frappe.ui.form.on("Delivery Note Item", "batch_no", function(frm, cdt, cdn) {
//     var d = locals[cdt][cdn];
//         frappe.db.get_value("Batch", {"name": d.batch_no}, "purchase_rate", function(value) {
//             d.purchase_rate = value.purchase_rate;
//         });
// });

File: jammy_app/jammy_app/custom_jammy/delivery_note/delivery_note.py
from __future__ import unicode_literals
from frappe.model.document import Document
import frappe
from frappe import utils
from frappe.utils.data import getdate, nowdate
from frappe.utils import get_link_to_form,flt
from six import string_types
from jammy_app.jammy_app.custom_jammy.quotation.quotation import set_total_cartons_and_total_weight
from itertools import groupby

def validate(doc, method):
	set_total_cartons_and_total_weight(doc)

@frappe.whitelist()
def create_bol(doctype,docname,type_of_bol):

	delivery_note = frappe.get_doc(doctype, docname)
	# Check if BOL already exists for the delivery note
	existing_bol = frappe.get_all('Bill Of Lading JI', filters={'ref': delivery_note.name}, limit=1)
	if existing_bol:
		frappe.throw('A Bill of Lading already exists for this Delivery Note.')

	bill_of_lading = frappe.new_doc('Bill Of Lading JI')
	
	bill_of_lading.ship_from = delivery_note.set_warehouse
	bill_of_lading.date = getdate(nowdate())
	bill_of_lading.ref = delivery_note.name
	bill_of_lading.carrier_name = delivery_note.ship_via
	bill_of_lading.shipping_address_name=delivery_note.shipping_address_name
	bill_of_lading.ship_to = delivery_note.shipping_address
	bill_of_lading.cid = delivery_note.po_no
	bill_of_lading.freight_charge_terms='Prepaid'
	bill_of_lading.third_party_freight_charges_bill_to=delivery_note.company_address_display
	total_weight = 0
	total_cartons = 0
	pallet_for_total_carton = frappe.db.get_single_value('Bill Of Lading Settings JI', 'pallet_for_total_cartons')
	weight_of_one_pallet = frappe.db.get_single_value('Bill Of Lading Settings JI', 'weight_of_one_pallet')

	
	if(type_of_bol=="direct_item_info"):
		bill_of_lading.type_of_bol='Direct DN Item info'
		for item in delivery_note.items:
			item_group = item.item_group
			nmfc_cf = frappe.get_value('Item Group', item_group, 'nmfc_cf')
			freight_class_cf = frappe.get_value('Item Group', item_group, 'freight_class_cf')
			description = f"{item.item_name}:{item.item_group}"
			weight = item.gross_weight
			total_weight += weight
			carton = item.cartons
			total_cartons += carton
			
			row = bill_of_lading.append('bill_of_lading_details',{
				"description":description,
				"weight": item.gross_weight,
				"packing_units": item.cartons,
				"nmfc": nmfc_cf,
				"freight_class":freight_class_cf,
			})
	
	elif(type_of_bol=="one_line_bol"):
		bill_of_lading.type_of_bol='One line BOL item'
		
		aggregated_weight = sum([float(item.gross_weight) for item in delivery_note.items])
		aggregated_cartons = sum([float(item.cartons) for item in delivery_note.items])
		total_weight=aggregated_weight
		total_cartons=aggregated_cartons
		one_line_bol_item_description = frappe.db.get_single_value('Bill Of Lading Settings JI', 'one_line_bol_item_description')
		one_line_bol_item_nmfc = frappe.db.get_single_value('Bill Of Lading Settings JI', 'one_line_bol_item_nmfc')
		one_line_bol_item_freight_class = frappe.db.get_single_value('Bill Of Lading Settings JI', 'one_line_bol_item_freight_class')

		row = bill_of_lading.append('bill_of_lading_details',{
				"description":one_line_bol_item_description,
				"weight": aggregated_weight,
				"packing_units": aggregated_cartons,
				"nmfc": one_line_bol_item_nmfc,
				"freight_class":one_line_bol_item_freight_class,
			})

	elif(type_of_bol=="normal_bol"):
		bill_of_lading.type_of_bol='Normal BOL(aggregate by item group)'
		bol_rows=[]
		for item in delivery_note.items:
			if not any(d['description'] == item.item_group for  d in bol_rows):
				nmfc_cf = frappe.get_value('Item Group', item.item_group, 'nmfc_cf')
				freight_class_cf = frappe.get_value('Item Group',item.item_group, 'freight_class_cf')	
				total_weight=total_weight+item.gross_weight		
				total_cartons=total_cartons+item.cartons
				bol_rows.append({'description':item.item_group,'weight':item.gross_weight,'packing_units':item.cartons,'nmfc':nmfc_cf,'freight_class':freight_class_cf})
			else:
				for d in bol_rows:
					if d['description'] ==item.item_group:
						d['weight']=d['weight']+item.gross_weight
						d['packing_units']=d['packing_units']+item.cartons
						total_weight=total_weight+item.gross_weight		
						total_cartons=total_cartons+item.cartons						
						break
		for row in bol_rows:
			bill_of_lading.append('bill_of_lading_details',row)

	bill_of_lading.total_cartons = total_cartons
	bill_of_lading.pallet_quantity = total_cartons % pallet_for_total_carton
	bill_of_lading.total_weight = total_weight+(35 * bill_of_lading.pallet_quantity)
	bill_of_lading.save(ignore_permissions=True)
	msg = ('Bill Of Lading {} is created'.format(frappe.bold(get_link_to_form('Bill Of Lading JI', bill_of_lading.name))))
	frappe.msgprint(msg)
